File: server/src/ssh/certificates.ts
import ssh2, { type ParsedKey } from 'ssh2';

const { utils } = ssh2;

const CERT_TYPE_RE =
  /^(ssh-(?:rsa|dss|ed25519)|ecdsa-sha2-nistp(?:256|384|521))-cert-v0[01]@openssh\.com$/;

export interface OpenSshCertificate {
  source: Buffer;
  type: string;
  baseType: string;
  publicBlob: Buffer;
  keyFields: Buffer[];
}

/** Parse an SSH public/private key with ssh2's supported key formats. */
export function parseSshKey(
  source: Buffer | string | ParsedKey,
  passphrase?: Buffer | string,
): ParsedKey | Error {
  return utils.parseKey(source, passphrase);
}

interface SshField {
  value: Buffer;
  next: number;
}

function readSshField(data: Buffer, offset: number): SshField | undefined {
  if (offset < 0 || offset + 4 > data.length) return undefined;
  const length = data.readUInt32BE(offset);
  const start = offset + 4;
  const end = start + length;
  if (end > data.length) return undefined;
  return { value: data.subarray(start, end), next: end };
}

function fieldCount(type: string): number | undefined {
  if (type === 'ssh-rsa') return 2;
  if (type === 'ssh-dss') return 4;
  if (type === 'ssh-ed25519') return 1;
  if (type.startsWith('ecdsa-sha2-')) return 2;
  return undefined;
}

function readKeyFields(
  blob: Buffer,
  expectedType: string,
  certificate: boolean,
): Buffer[] | undefined {
  const type = readSshField(blob, 0);
  if (!type || type.value.toString() !== expectedType) return undefined;
  let offset = type.next;
  if (certificate) {
    const nonce = readSshField(blob, offset);
    if (!nonce) return undefined;
    offset = nonce.next;
  }
  const count = fieldCount(
    certificate ? expectedType.replace(/-cert-v0[01]@openssh\.com$/, '') : expectedType,
  );
  if (count === undefined) return undefined;
  const fields: Buffer[] = [];
  for (let i = 0; i < count; i++) {
    const field = readSshField(blob, offset);
    if (!field) return undefined;
    fields.push(field.value);
    offset = field.next;
  }
  return fields;
}

/** Parse the public line written by ssh-keygen for an OpenSSH user certificate. */
export function parseOpenSshCertificate(
  source: Buffer | string,
): OpenSshCertificate | Error {
  const bytes = Buffer.isBuffer(source) ? source : Buffer.from(source);
  const line = bytes.toString('utf8').trim();
  const match = /^(\S+)\s+([A-Za-z0-9+/]+={0,2})(?:\s|$)/.exec(line);
  if (!match) return new Error('unsupported certificate format');
  const type = match[1] ?? '';
  const typeMatch = CERT_TYPE_RE.exec(type);
  if (!typeMatch) return new Error(`unsupported certificate type: ${type || 'unknown'}`);
  const publicBlob = Buffer.from(match[2] ?? '', 'base64');
  const keyFields = readKeyFields(publicBlob, type, true);
  if (!keyFields) return new Error('malformed OpenSSH certificate');

  // ssh2 recognizes certificate objects internally, which lets them flow
  // through authHandler. Its public parser does not retain the complete
  // certificate blob, so the wrapper below replaces that accessor.
  const parsed = parseSshKey(bytes);
  if (parsed instanceof Error) return parsed;

  return {
    source: bytes,
    type,
    baseType: typeMatch[1] ?? '',
    publicBlob,
    keyFields,
  };
}

/** True when a certificate contains the public half of this private key. */
export function certificateMatchesKey(
  certificate: OpenSshCertificate,
  privateKey: ParsedKey,
): boolean {
  const fields = readKeyFields(
    privateKey.getPublicSSH(),
    certificate.baseType,
    false,
  );
  return (
    fields?.length === certificate.keyFields.length
    && fields.every((field, i) => field.equals(certificate.keyFields[i]!))
  );
}

/** Authentication algorithm variants to try for this certificate. */
export function certificateAlgorithms(certificate: OpenSshCertificate): string[] {
  if (certificate.baseType !== 'ssh-rsa') return [certificate.type];
  const suffix = certificate.type.slice('ssh-rsa'.length);
  // Prefer SHA-2, as current OpenSSH disables ssh-rsa/SHA-1 by default, while
  // retaining the SHA-1 variant for older servers.
  return [
    `rsa-sha2-512${suffix}`,
    `rsa-sha2-256${suffix}`,
    certificate.type,
  ];
}

interface DerValue {
  value: Buffer;
  next: number;
}

function readDerValue(data: Buffer, offset: number, tag: number): DerValue | undefined {
  if (data[offset] !== tag || offset + 2 > data.length) return undefined;
  let length = data[offset + 1] ?? 0;
  let start = offset + 2;
  if ((length & 0x80) !== 0) {
    const width = length & 0x7f;
    if (width === 0 || width > 4 || start + width > data.length) return undefined;
    length = 0;
    for (let i = 0; i < width; i++) length = (length * 256) + (data[start + i] ?? 0);
    start += width;
  }
  const end = start + length;
  if (end > data.length) return undefined;
  return { value: data.subarray(start, end), next: end };
}

function sshField(value: Buffer): Buffer {
  const out = Buffer.allocUnsafe(4 + value.length);
  out.writeUInt32BE(value.length, 0);
  value.copy(out, 4);
  return out;
}

/** Convert OpenSSL's DER ECDSA signature into SSH's pair of mpints. */
function ecdsaSignature(signature: Buffer): Buffer | Error {
  const sequence = readDerValue(signature, 0, 0x30);
  if (!sequence) return new Error('invalid ECDSA signature');
  const r = readDerValue(sequence.value, 0, 0x02);
  const s = r && readDerValue(sequence.value, r.next, 0x02);
  if (!r || !s || s.next !== sequence.value.length) {
    return new Error('invalid ECDSA signature');
  }
  return Buffer.concat([sshField(r.value), sshField(s.value)]);
}

function dsaSignature(signature: Buffer): Buffer | Error {
  if (signature.length === 40) return signature;
  const sequence = readDerValue(signature, 0, 0x30);
  if (!sequence) return new Error('invalid DSA signature');
  const r = readDerValue(sequence.value, 0, 0x02);
  const s = r && readDerValue(sequence.value, r.next, 0x02);
  if (!r || !s) return new Error('invalid DSA signature');
  const out = Buffer.alloc(40);
  r.value.subarray(Math.max(0, r.value.length - 20)).copy(
    out,
    Math.max(0, 20 - r.value.length),
  );
  s.value.subarray(Math.max(0, s.value.length - 20)).copy(
    out,
    20 + Math.max(0, 20 - s.value.length),
  );
  return out;
}

/**
 * Combine a public OpenSSH certificate with its private IdentityFile key.
 * ssh2 has no CertificateFile option, but its auth handler accepts a parsed
 * key object. Keeping ssh2's internal parsed-key marker while replacing the
 * public blob and signer gives it the exact certificate authentication shape.
 */
export function certifiedKey(
  privateKey: ParsedKey,
  certificate: OpenSshCertificate,
  algorithm: string,
): ParsedKey | Error {
  if (!certificateMatchesKey(certificate, privateKey)) {
    return new Error('certificate does not match private key');
  }
  if (!certificateAlgorithms(certificate).includes(algorithm)) {
    return new Error(`unsupported certificate algorithm: ${algorithm}`);
  }
  const parsed = parseSshKey(certificate.source);
  if (parsed instanceof Error) return parsed;

  const hash =
    algorithm.startsWith('rsa-sha2-512') ? 'sha512'
    : algorithm.startsWith('rsa-sha2-256') ? 'sha256'
    : algorithm.startsWith('ssh-rsa-') ? 'sha1'
    : undefined;
  const mutable = parsed as unknown as {
    type: string;
    sign(data: Buffer | string): Buffer | Error;
    verify: ParsedKey['verify'];
    isPrivateKey(): boolean;
    getPrivatePEM: ParsedKey['getPrivatePEM'];
    getPublicPEM: ParsedKey['getPublicPEM'];
    getPublicSSH(): Buffer;
    equals: ParsedKey['equals'];
  };
  mutable.type = algorithm;
  mutable.sign = (data) => {
    const signature = privateKey.sign(data, hash);
    if (certificate.baseType.startsWith('ecdsa-sha2-')) {
      return ecdsaSignature(signature);
    }
    if (certificate.baseType === 'ssh-dss') return dsaSignature(signature);
    return signature;
  };
  mutable.verify = privateKey.verify.bind(privateKey);
  mutable.isPrivateKey = () => true;
  mutable.getPrivatePEM = privateKey.getPrivatePEM.bind(privateKey);
  mutable.getPublicPEM = privateKey.getPublicPEM.bind(privateKey);
  mutable.getPublicSSH = () => certificate.publicBlob;
  mutable.equals = privateKey.equals.bind(privateKey);
  return parsed;
}

File: app/internal/sshx/certificates.go
package sshx

// OpenSSH user certificate handling: parse the public line written by
// ssh-keygen (the CertificateFile), pair it with its IdentityFile private
// key, and produce a signer that authenticates with the certificate.
//
// Port of server/src/ssh/certificates.ts. The ssh2-specific plumbing there
// (DER signature re-encoding, ParsedKey mutation) has no Go counterpart:
// x/crypto/ssh emits SSH wire-format signatures natively and provides
// ssh.NewCertSigner for the certificate authentication shape.

import (
	"bytes"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"fmt"
	"regexp"
	"slices"
	"strings"

	"golang.org/x/crypto/ssh"
)

var (
	certTypeRe   = regexp.MustCompile(`^(ssh-(?:rsa|dss|ed25519)|ecdsa-sha2-nistp(?:256|384|521))-cert-v0[01]@openssh\.com$`)
	certLineRe   = regexp.MustCompile(`^(\S+)\s+([A-Za-z0-9+/]+={0,2})(?:\s|$)`)
	certSuffixRe = regexp.MustCompile(`-cert-v0[01]@openssh\.com$`)
)

// OpenSshCertificate mirrors the OpenSshCertificate shape of the TypeScript
// module, plus the parsed x/crypto certificate for signer construction.
type OpenSshCertificate struct {
	Source      []byte
	Type        string
	BaseType    string
	PublicBlob  []byte
	KeyFields   [][]byte
	Certificate *ssh.Certificate
}

// ParseSshPrivateKey parses an SSH private key, mirroring parseSshKey: with a
// passphrase the encrypted form is decrypted, otherwise the key must be plain.
func ParseSshPrivateKey(source []byte, passphrase []byte) (ssh.Signer, error) {
	if len(passphrase) > 0 {
		return ssh.ParsePrivateKeyWithPassphrase(source, passphrase)
	}
	return ssh.ParsePrivateKey(source)
}

// certReadField reads one uint32-length-prefixed SSH field.
func certReadField(data []byte, offset int) (value []byte, next int, ok bool) {
	if offset < 0 || offset+4 > len(data) {
		return nil, 0, false
	}
	length := int(binary.BigEndian.Uint32(data[offset:]))
	start := offset + 4
	end := start + length
	if length < 0 || end > len(data) {
		return nil, 0, false
	}
	return data[start:end], end, true
}

func certFieldCount(keyType string) int {
	switch {
	case keyType == "ssh-rsa":
		return 2
	case keyType == "ssh-dss":
		return 4
	case keyType == "ssh-ed25519":
		return 1
	case strings.HasPrefix(keyType, "ecdsa-sha2-"):
		return 2
	}
	return -1
}

func certReadKeyFields(blob []byte, expectedType string, certificate bool) ([][]byte, bool) {
	typ, offset, ok := certReadField(blob, 0)
	if !ok || string(typ) != expectedType {
		return nil, false
	}
	if certificate {
		_, next, ok := certReadField(blob, offset) // nonce
		if !ok {
			return nil, false
		}
		offset = next
	}
	baseType := expectedType
	if certificate {
		baseType = certSuffixRe.ReplaceAllString(expectedType, "")
	}
	count := certFieldCount(baseType)
	if count < 0 {
		return nil, false
	}
	fields := make([][]byte, 0, count)
	for i := 0; i < count; i++ {
		field, next, ok := certReadField(blob, offset)
		if !ok {
			return nil, false
		}
		fields = append(fields, field)
		offset = next
	}
	return fields, true
}

// ParseOpenSshCertificate parses the public line written by ssh-keygen for an
// OpenSSH user certificate.
func ParseOpenSshCertificate(source []byte) (*OpenSshCertificate, error) {
	line := strings.TrimSpace(string(source))
	match := certLineRe.FindStringSubmatch(line)
	if match == nil {
		return nil, errors.New("unsupported certificate format")
	}
	certType := match[1]
	typeMatch := certTypeRe.FindStringSubmatch(certType)
	if typeMatch == nil {
		name := certType
		if name == "" {
			name = "unknown"
		}
		return nil, fmt.Errorf("unsupported certificate type: %s", name)
	}
	publicBlob, err := base64.StdEncoding.DecodeString(match[2])
	if err != nil {
		return nil, errors.New("malformed OpenSSH certificate")
	}
	keyFields, ok := certReadKeyFields(publicBlob, certType, true)
	if !ok {
		return nil, errors.New("malformed OpenSSH certificate")
	}

	parsed, err := ssh.ParsePublicKey(publicBlob)
	if err != nil {
		return nil, err
	}
	cert, ok := parsed.(*ssh.Certificate)
	if !ok {
		return nil, errors.New("malformed OpenSSH certificate")
	}

	return &OpenSshCertificate{
		Source:      bytes.Clone(source),
		Type:        certType,
		BaseType:    typeMatch[1],
		PublicBlob:  publicBlob,
		KeyFields:   keyFields,
		Certificate: cert,
	}, nil
}

// CertificateMatchesKey reports whether the certificate contains the public
// half of this key.
func CertificateMatchesKey(certificate *OpenSshCertificate, publicKey ssh.PublicKey) bool {
	fields, ok := certReadKeyFields(publicKey.Marshal(), certificate.BaseType, false)
	if !ok || len(fields) != len(certificate.KeyFields) {
		return false
	}
	for i, field := range fields {
		if !bytes.Equal(field, certificate.KeyFields[i]) {
			return false
		}
	}
	return true
}

// CertificateAlgorithms lists the authentication algorithm variants to try
// for this certificate.
func CertificateAlgorithms(certificate *OpenSshCertificate) []string {
	if certificate.BaseType != "ssh-rsa" {
		return []string{certificate.Type}
	}
	suffix := certificate.Type[len("ssh-rsa"):]
	// Prefer SHA-2, as current OpenSSH disables ssh-rsa/SHA-1 by default,
	// while retaining the SHA-1 variant for older servers.
	return []string{
		"rsa-sha2-512" + suffix,
		"rsa-sha2-256" + suffix,
		certificate.Type,
	}
}

// CertifiedKey combines a public OpenSSH certificate with its private
// IdentityFile key into a signer whose public key is the certificate, giving
// the exact certificate authentication shape.
func CertifiedKey(privateKey ssh.Signer, certificate *OpenSshCertificate, algorithm string) (ssh.Signer, error) {
	if !CertificateMatchesKey(certificate, privateKey.PublicKey()) {
		return nil, errors.New("certificate does not match private key")
	}
	if !slices.Contains(CertificateAlgorithms(certificate), algorithm) {
		return nil, fmt.Errorf("unsupported certificate algorithm: %s", algorithm)
	}

	// For the rsa-sha2-* certificate variants, pin the underlying signature
	// algorithm the way the TypeScript wrapper picks the digest.
	signKey := privateKey
	if base := certSuffixRe.ReplaceAllString(algorithm, ""); base != privateKey.PublicKey().Type() {
		if algoSigner, ok := privateKey.(ssh.AlgorithmSigner); ok {
			if pinned, err := ssh.NewSignerWithAlgorithms(algoSigner, []string{base}); err == nil {
				signKey = pinned
			}
		}
	}
	return ssh.NewCertSigner(certificate.Certificate, signKey)
}
